lib/dictBuilder/cover.c

```diff
@@ -969,21 +1074,22 @@ static void COVER_tryParameters(void *opaque) {
   {
     const size_t tail = COVER_buildDictionary(ctx, freqs, &activeDmers, dict,
                                               dictBufferCapacity, parameters);
-    dictBufferCapacity = ZDICT_finalizeDictionary(
-        dict, dictBufferCapacity, dict + tail, dictBufferCapacity - tail,
-        ctx->samples, ctx->samplesSizes, (unsigned)ctx->nbTrainSamples,
-        parameters.zParams);
-    if (ZDICT_isError(dictBufferCapacity)) {
-      DISPLAYLEVEL(1, "Failed to finalize dictionary\n");
+    const COVER_dictSelection_t selection = COVER_selectDict(dict + tail, dictBufferCapacity - tail,
+        ctx->samples, ctx->samplesSizes, (unsigned)ctx->nbTrainSamples, ctx->nbTrainSamples, ctx->nbSamples, parameters, ctx->offsets,
+        totalCompressedSize);
+
+    if (COVER_dictSelectionIsError(selection)) {
+      DISPLAYLEVEL(1, "Failed to select dictionary\n");
+      dictBufferCapacity = selection.totalCompressedSize;
+      totalCompressedSize = selection.totalCompressedSize;
       goto _cleanup;
     }
-  }
-  /* Check total compressed size */
-  totalCompressedSize = COVER_checkTotalCompressedSize(parameters, ctx->samplesSizes,
-                                                       ctx->samples, ctx->offsets,
-                                                       ctx->nbTrainSamples, ctx->nbSamples,
-                                                       dict, dictBufferCapacity);
+    dictBufferCapacity = selection.dictSize;
+    totalCompressedSize = selection.totalCompressedSize;
+    memcpy(dict, selection.dictContent, dictBufferCapacity);
```

Comment:
We no longer need `dict`. If we move the declaration of `selection` to the top scope, we can refactor `COVER_best_finish()` to take a `COVER_dictSelection` instead of the three parameters (dict, totalCompressedSize, dictBufferCapacity).


```c
COVER_dictSelection_t selection = COVER_dictSelectionError(ERROR(GENERIC));
/* ... */
{
  const size_t tail = COVER_buildDictionary(/* ... */);
  selection = COVER_selectDict(/* ... */);
}
_cleanup:
  COVER_best_finish(data->best, parameters, selection); /* Copies the dict out of selection */
  COVER_dictSelectionFree(selection);

``